apps/sel4test-tests/src/tests/cache.c

```diff
@@ -327,3 +327,229 @@ static int test_page_uncached_after_retype(env_t env)
 DEFINE_TEST(CACHEFLUSH0004, "Test that mapping a frame uncached doesn't see stale data after retype",
             test_page_uncached_after_retype,
             config_set(CONFIG_HAVE_CACHE))
+
+#if defined(CONFIG_ARCH_ARM)
+
+/*
+ * The following tests aim to test the seL4_ARM_* cache operations on every
+ * kind of frame mapping that can be created. The motivation behind this is
+ * that some ARM hardware cache instructions generate faults depending on the
+ * permissions of the mapping and we need to ensure that the kernel catches
+ * these so that a fault does not occur in kernel space. In addition, these
+ * tests ensure that the kernel is enforcing a frame's cap rights, depending
+ * on the cache operation.
+ *
+ * For each kind of mapping we need to test the following operations:
+ *
+ * seL4_ARM_VSpace_Clean_Data
+ * seL4_ARM_VSpace_Invalidate_Data
+ * seL4_ARM_VSpace_CleanInvalidate_Data
+ * seL4_ARM_VSpace_Unify_Instruction
+ * seL4_ARM_Page_Clean_Data
+ * seL4_ARM_Page_Invalidate_Data
+ * seL4_ARM_Page_CleanInvalidate_Data
+ * seL4_ARM_Page_Unify_Instruction
+ *
+ */
+static int test_cache_invalid(env_t env)
+{
+    seL4_CPtr frame;
+    uintptr_t vstart;
+    int err;
+    void *vaddr;
+    reservation_t reservation;
+    vka_t *vka = &env->vka;
+
+    reservation = vspace_reserve_range(&env->vspace, PAGE_SIZE_4K, seL4_AllRights, 1, &vaddr);
+    assert(reservation.res);
+
+    vstart = (uintptr_t)vaddr;
+    assert(IS_ALIGNED(vstart, seL4_PageBits));
+
+    /* Create a frame, but deliberately do not create a mapping for it. */
+    frame = vka_alloc_frame_leaky(vka, PAGE_BITS_4K);
+    test_assert(frame != seL4_CapNull);
```

Comment:
I'd move this into a separate function which with as parameter the rights to pass to `vspace_reserve_range` and call it in all the tests. That would save a lot of code duplication and make it easier to see the bits that change between the tests. At least I think it's all identical otherwise.

If not doing this, at least make all the code exactly the same without subtle differences for no reason, so it's easier to see it's identical.